File: Tic Tac Toe/src/TicTacToe.cpp
#include <iostream>
#include <string>
#include <iomanip>
#include "TicTacToe.h"

bool inlimit(int since, int until, int input)
{
	return (input >= since && input <= until) ? true : false;
}

void toggle(bool& boolean)
{
	if (boolean)
		boolean = false;
	else boolean = true;
}

TicTacToe::TicTacToe(int board_width)
	: board_width(board_width), board_area(board_width * board_width), end(false), round(1), hide(false), winner("NULL")
{
	tile_display = new int*[board_width]; // Dynamic Allocation for the board
	for (int row = 0; row < board_width; row++)
		tile_display[row] = new int[board_width];

	for (int row = 0; row < board_width; row++)
		for (int column = 0; column < board_width; column++)// set all tiles on the board with numbers
			tile_display[row][column] = column + row * board_width + 1;

	tile_raw = new char*[board_width]; // Dynamic Allocation for the board
	for (int row = 0; row < board_width; row++)
		tile_raw[row] = new char[board_width];

	for (int row = 0; row < board_width; row++)
		for (int column = 0; column < board_width; column++)// set all tiles on the board with numbers
			tile_raw[row][column] = '\0';
}

TicTacToe::~TicTacToe()
{
	for (int i = 0; i < board_width; i++)
		delete[] tile_display[i];
	delete[] tile_display;
}

void TicTacToe::setPlayer1_name(std::string name)
{
	player1_name = name;
}

void TicTacToe::setPlayer2_name(std::string name)
{
	player2_name = name;
}

void TicTacToe::start()
{
	
	//// round amount
	//std::cout << "Enter the amount of round(s) >> ";
	//int input;
	//while (true) {
	//	std::cin >> input;
	//	if (inlimit(1, 10, round)) {
	//		round = input;
	//		break;
	//	}
	//	else std::cout << "The amount of round must be under 1 to 10, please retry >> ";
	//}
	
	// player names
	std::cout << "Please enter the name of Player #1 >> ";
	std::string input;
	std::cin >> input;
	setPlayer1_name(input);
	std::cout << "Please enter the name of Player #2 >> ";
	std::cin >> input;
	setPlayer2_name(input);
	std::cout << "\n-------------------------------------------\n";

	
}

void TicTacToe::run()
{
	std::string ask_message = "Answer";
	while (!end) {
		std::cout << std::endl << "O : " << player1_name << std::endl << "X : " << player2_name << std::endl << std::endl;
		drawboard();
		std::cout << "Current turn : " << (round % 2 ? player1_name : player2_name) << (round % 2 ? " [O]" : " [X]")  << std::endl << std::endl;
		std::cout << ask_message << " >";
		std::cin >> last_answer;
		currentrow = (last_answer - 1) / board_width;
		currentcolumn = (last_answer - 1) % board_width;
		if (last_answer == 0) {
			toggle(hide);
			system("cls");
			ask_message = "Answer";
			continue;
		}
		else if (!inlimit(0, board_area, last_answer)) {
			ask_message = "Tile not found, please retry";
		}
		else if (tile_raw[(last_answer - 1) / board_width][(last_answer - 1) % board_width] != '\0' && last_answer != 0) {
			ask_message = "Tile not empty, please retry";
		}
		else { // Pass
			ask_message = "Answer";
			tile_raw[(last_answer - 1) / board_width][(last_answer - 1) % board_width] = (round % 2 ? 'O' : 'X');
			round++;
		}

		

		checkwin();
		system("cls");


	}
}

void TicTacToe::drawboard()
{

	for (int i = 0; i < board_width * 6 + 1; i++)
		std::cout << "_";
	for (int row = 0; row < board_width; row++) {
		std::cout << std::endl;
		std::cout << "|";
		for (int i = 0; i < board_width; i++)
			std::cout << "     |";
		std::cout << std::endl << "| ";
		for (int column = 0; column < board_width; column++) {
			if (tile_raw[row][column] == 0)
				std::cout << std::setw(3) << (hide ? (std::string)" " : std::to_string(tile_display[row][column]));
			else
				std::cout << " " << tile_raw[row][column] << " ";
			std::cout << " | ";
		}
		std::cout << std::endl;
		std::cout << "|";
		for (int i = 0; i < board_width; i++)
			std::cout << "_____|";
	}
	std::cout << std::endl << std::endl;
}

void TicTacToe::checkwin()
{
	bool havewin = false;
	int th, tv, tdl = 0, tdr = 0;
	for (int i = 0; i < board_width && !havewin; i++) { // For player 1
		th = 0; tv = 0;
		for (int j = 0; j < board_width && !havewin; j++) {

			if (tile_raw[i][j] != 'O') { // Horizontal
				th = 0;
			}
			else if (th == board_width - 1) {
				havewin = true;
				winner = player1_name;
				end = true;
			}
			else th++;
			if (tile_raw[j][i] != 'O') { // Vertical
				tv = 0;
			}
			else if (tv == board_width - 1) {
				havewin = true;
				winner = player1_name;
				end = true;
			}
			else tv++;



		}
		if (tile_raw[i][i] != 'O') { // Diagonal
			tdl = 0;
		}
		else if (tdl == board_width - 1) {
			havewin = true;
			winner = player1_name;
			end = true;
		}
		else tdl++;
		if (tile_raw[i][board_width - i - 1] != 'O') { // Diagonal
			tdr = 0;
		}
		else if (tdr == board_width - 1) {
			havewin = true;
			winner = player1_name;
			end = true;
		}
		else tdr++;
	}

	th = 0; tv = 0; tdl = 0; tdr = 0;
	for (int i = 0; i < board_width && !havewin; i++) { // For player 2
		th = 0; tv = 0;
		for (int j = 0; j < board_width && !havewin; j++) {

			if (tile_raw[i][j] != 'X') { // Horizontal
				th = 0;
			}
			else if (th == board_width - 1) {
				havewin = true;
				winner = player2_name;
				end = true;
			}
			else th++;
			if (tile_raw[j][i] != 'X') { // Vertical
				tv = 0;
			}
			else if (tv == board_width - 1) {
				havewin = true;
				winner = player2_name;
				end = true;
			}
			else tv++;



		}
		if (tile_raw[i][i] != 'X') { // Diagonal
			tdl = 0;
		}
		else if (tdl == board_width - 1) {
			havewin = true;
			winner = player2_name;
			end = true;
		}
		else tdl++;
		if (tile_raw[i][board_width - i - 1] != 'X') { // Diagonal
			tdr = 0;
		}
		else if (tdr == board_width - 1) {
			havewin = true;
			winner = player2_name;
			end = true;
		}
		else tdr++;
	}
	if (round == board_area + 1)
		end = true;
}

void TicTacToe::summary()
{
	hide = true;
	drawboard();
	std::cout << std::endl;
	if (winner != "NULL") {
		std::cout << "Winner : " << (winner == player1_name ? player1_name : player2_name) << (winner == player1_name ? " [O]" : " [X]") << " !";
	}
	else {
		std::cout << "DRAW !";
	}
}

File: javascript/tictactoe.js
class TicTacToe {

    constructor(board_size, screen_width) {
        this.board_width = board_size;
        this.board_area = board_size * board_size;
        this.end = false;
        this.winner = null; //1: x 2: o
        this.currentrow;
        this.currentcolumn;
        this.tile_width = screen_width / this.board_width;
        this.round = 0;
        this.win_type = 0; //1:hori 2:vert 3:diag\ 4:diag/
        this.win_at = null;

        this.tile_raw = []; // Dynamic Allocation for the board
        for (let row = 0; row < this.board_width; row++)
            this.tile_raw[row] = [];

        for (let row = 0; row < this.board_width; row++)
            for (let column = 0; column < this.board_width; column++) // set all tiles on the board with numbers
                this.tile_raw[row][column] = null;
    }

    setState(mX, mY) {
        if (this.winner == null) {
            var x = null;
            var y = null;
            for (let i = 1; i <= this.board_width; i++) {
                if (x == null && mX < this.tile_width * i) {
                    x = i - 1;
                }
                if (y == null && mY < this.tile_width * i) {
                    y = i - 1;
                }
            }
            if (this.tile_raw[x][y] == null) {
                this.tile_raw[x][y] = this.round % 2 + 1;
                this.round++;
            }
        }
    }

    drawboard() {
        for (let i = 1; i < this.board_width; i++) {
            strokeWeight(2);
            stroke(0);
            noFill();
            line(this.tile_width * i, 0, this.tile_width * i, this.tile_width * this.board_width);
            line(0, this.tile_width * i, this.tile_width * this.board_width, this.tile_width * i);
        }
        for (let i = 0; i < this.board_width; i++) {
            for (let j = 0; j < this.board_width; j++) {
                if (this.tile_raw[i][j] == 1) {
                    strokeWeight(2);
                    stroke(255, 0, 0);
                    this.drawCross(this.tile_width * i + this.tile_width / 2, this.tile_width * j + this.tile_width / 2, this.tile_width / 4);
                } else if (this.tile_raw[i][j] == 2) {
                    strokeWeight(2);
                    stroke(0, 0, 255);
                    this.drawO(this.tile_width * i + this.tile_width / 2, this.tile_width * j + this.tile_width / 2, this.tile_width / 4);
                }
            }
        }
        if (this.winner != null) {
            this.drawWinLine();
        }
    }

    drawCross(midX, midY, radius) {
        line(midX - radius, midY - radius, midX + radius, midY + radius);
        line(midX + radius, midY - radius, midX - radius, midY + radius);
    }

    drawO(x, y, radius) {
        //ellipse(x - radius, y - radius, radius * 2, radius * 2);
        ellipse(x, y, radius * 2, radius * 2);
    }

    drawWinLine() {
        var m = 6;
        strokeWeight(3);
        if (this.winner == 2) stroke(0, 0, 255);
        else stroke(255, 0, 0);
        switch (this.win_type) {
            case 2: //horizontal
                line(this.tile_width / m, this.win_at * this.tile_width + this.tile_width / 2, this.board_width * this.tile_width - this.tile_width / m, this.win_at * this.tile_width + this.tile_width / 2);
                break;
            case 1: //vertical
                line(this.win_at * this.tile_width + this.tile_width / 2, this.tile_width / m, this.win_at * this.tile_width + this.tile_width / 2, this.board_width * this.tile_width - this.tile_width / m);
                break;
            case 3: //diagonal \
                line(this.tile_width / m, this.tile_width / m, this.board_width * this.tile_width - this.tile_width / m, this.board_width * this.tile_width - this.tile_width / m);
                break;
            case 4: //diagonal /
                line(this.tile_width / m, this.board_width * this.tile_width - this.tile_width / m, this.board_width * this.tile_width - this.tile_width / m, this.tile_width / m);
                break;
        }
    }

    checkWin() {
        var havewin = false;
        var th, tv, tdl = 0,
            tdr = 0;
        for (let i = 0; i < this.board_width && !havewin; i++) { // For player 1
            th = 0;
            tv = 0;
            for (let j = 0; j < this.board_width && !havewin; j++) {

                if (this.tile_raw[i][j] != 1) { // Horizontal
                    th = 0;
                } else if (th == this.board_width - 1) {
                    havewin = true;
                    this.win_type = 1;
                    this.win_at = i;
                    this.winner = 1;
                    this.end = true;
                } else th++;
                if (this.tile_raw[j][i] != 1) { // Vertical
                    tv = 0;
                } else if (tv == this.board_width - 1) {
                    havewin = true;
                    this.win_type = 2;
                    this.win_at = i;
                    this.winner = 1;
                    this.end = true;
                } else tv++;
            }
            if (this.tile_raw[i][i] != 1) { // Diagonal
                tdl = 0;
            } else if (tdl == this.board_width - 1) {
                havewin = true;
                this.win_type = 3;
                this.win_at = i;
                this.winner = 1;
                this.end = true;
            } else tdl++;
            if (this.tile_raw[i][this.board_width - i - 1] != 1) { // Diagonal
                tdr = 0;
            } else if (tdr == this.board_width - 1) {
                havewin = true;
                this.win_type = 4;
                this.win_at = i;
                this.winner = 1;
                this.end = true;
            } else tdr++;
        }

        th = 0;
        tv = 0;
        tdl = 0;
        tdr = 0;
        for (let i = 0; i < this.board_width && !havewin; i++) { // For player 2
            th = 0;
            tv = 0;
            for (let j = 0; j < this.board_width && !havewin; j++) {

                if (this.tile_raw[i][j] != 2) { // Horizontal
                    th = 0;
                } else if (th == this.board_width - 1) {
                    havewin = true;
                    this.win_type = 1;
                    this.win_at = i;
                    this.winner = 2;
                    this.end = true;
                } else th++;
                if (this.tile_raw[j][i] != 2) { // Vertical
                    tv = 0;
                } else if (tv == this.board_width - 1) {
                    havewin = true;
                    this.win_type = 2;
                    this.win_at = i;
                    this.winner = 2;
                    this.end = true;
                } else tv++;
            }
            if (this.tile_raw[i][i] != 2) { // Diagonal
                tdl = 0;
            } else if (tdl == this.board_width - 1) {
                havewin = true;
                this.win_type = 3;
                this.win_at = i;
                this.winner = 2;
                this.end = true;
            } else tdl++;
            if (this.tile_raw[i][this.board_width - i - 1] != 2) { // Diagonal
                tdr = 0;
            } else if (tdr == this.board_width - 1) {
                havewin = true;
                this.win_type = 4;
                this.win_at = i;
                this.winner = 2;
                this.end = true;
            } else tdr++;
        }
        if (round == this.board_area + 1)
            this.end = true;
    }
}